File: app/src/main/java/com/sup/dev/android/views/elements/cards/CardMenu.java
package com.sup.dev.android.views.elements.cards;

import android.support.annotation.StringRes;
import android.view.View;
import android.widget.TextView;

import com.sup.dev.android.androiddevsup.R;
import com.sup.dev.android.app.SupAndroid;
import com.sup.dev.android.utils.interfaces.UtilsResources;
import com.sup.dev.java.classes.callbacks.simple.Callback;

public class CardMenu extends Card {

    private Callback onClick;
    private boolean dividerVisible = false;
    private boolean enabled = true;
    private int background = 0x01FF0000;

    private final UtilsResources utilsResources = SupAndroid.di.utilsResources();

    public Object tag;
    private String text;
    private boolean customColor;
    private int textColor;

    @Override
    public int getLayout() {
        return R.layout.card_menu;
    }

    @Override
    public void bindView(View view) {
        View vTouch = view.findViewById(R.id.touch);
        View vDivider = view.findViewById(R.id.divider);
        TextView vText =  view.findViewById(R.id.text);

        vDivider.setVisibility(dividerVisible ? View.VISIBLE : View.GONE);
        vTouch.setFocusable(onClick != null && enabled);
        vTouch.setClickable(onClick != null && enabled);
        vTouch.setEnabled(onClick != null && enabled);
        vTouch.setOnClickListener((onClick != null && enabled)?v -> onClick.callback():null);
        if (background != 0x01FF0000) view.setBackgroundColor(background);

        vText.setText(text);
        vText.setEnabled(enabled);
        if (customColor) vText.setTextColor(textColor);
    }

    //
    //  Setters
    //

    public CardMenu setOnClick(Callback onClick) {
        this.onClick = onClick;
        update();
        return this;
    }

    public CardMenu setDividerVisible(boolean dividerVisible) {
        this.dividerVisible = dividerVisible;
        update();
        return this;
    }

    public CardMenu setEnabled(boolean enabled) {
        this.enabled = enabled;
        update();
        return this;
    }

    public CardMenu setBackground(int background) {
        this.background = background;
        update();
        return this;
    }

    public CardMenu setText(@StringRes int text) {
        return setText(utilsResources.getString(text));
    }

    public CardMenu setText(String text) {
        this.text = text;
        update();
        return this;
    }

    public CardMenu setTextColor(int color) {
        customColor = true;
        textColor = color;
        update();
        return this;
    }

    //
    //  Getters
    //

    public String getText() {
        return text;
    }

    public boolean isEnabled() {
        return enabled;
    }
}

File: src/main/java/com/sup/dev/android/views/cards/CardMenu.kt
package com.sup.dev.android.views.cards

import android.graphics.Color
import android.graphics.drawable.Drawable
import androidx.annotation.ColorRes
import androidx.annotation.StringRes
import android.view.View
import android.widget.TextView
import com.sup.dev.android.R
import com.sup.dev.android.tools.ToolsResources
import com.sup.dev.android.tools.ToolsView
import com.sup.dev.android.views.views.ViewChip
import com.sup.dev.android.views.views.ViewIcon
import com.sup.dev.java.libs.debug.Debug

open class CardMenu(
        layout: Int = 0
) : Card(if (layout > 0) layout else R.layout.card_menu) {

    var onClick: ((View, Int, Int) -> Unit)? = null
    var onLongClick: ((View) -> Unit)? = null
    var dividerVisible = false
    var enabled = true
    var background = 0
    var visible = true

    var text = ""
    var chipText = ""
    var description = ""
    var customColor = false
    var textColor = 0
    var icon = 0
    var iconDrawable: Drawable? = null
    var iconFilter: Int? = null

    override fun bindView(view: View) {
        super.bindView(view)
        val vTouch:View? = view.findViewById(R.id.vTouch)
        val vChip:ViewChip? = view.findViewById(R.id.vChip)
        val vDivider:View? = view.findViewById(R.id.vDivider)
        val vText:TextView? = view.findViewById(R.id.vText)
        val vDescription:TextView? = view.findViewById(R.id.vDesc)
        val vIcon:ViewIcon? = view.findViewById(R.id.vIcon)

        if(vTouch != null) {
            vTouch.visibility = if (visible) View.VISIBLE else View.GONE
            vTouch.isFocusable = onClick != null && enabled
            vTouch.isClickable = onClick != null && enabled
            vTouch.isEnabled = onClick != null && enabled
            ToolsView.setOnClickCoordinates(vTouch) { v, x, y -> onClick(v,x,y) }
            if(onLongClick != null) vTouch.setOnLongClickListener { onLongClick!!.invoke(it); return@setOnLongClickListener true }
        }

        if(vChip != null){
            vChip.setBackground(ToolsResources.getAccentColor(view.context))
            vChip.text = chipText
        }

        if(vIcon != null) {
            if (iconDrawable != null) vIcon.setImageDrawable(iconDrawable)
            if (iconFilter != null) vIcon.setFilter(iconFilter!!)
            else if (icon == 0) vIcon.visibility = View.GONE
            else vIcon.setImageResource(icon)
        }

        if(vDivider != null) {
            vDivider.visibility = if (visible && dividerVisible) View.VISIBLE else View.GONE
        }

        if(vDescription != null) {
            vDescription.text = description
            vDescription.isEnabled = enabled
            vDescription.visibility = if (description.isNotEmpty()) View.VISIBLE else View.GONE
        }

        if(vText != null) {
            vText.text = text
            vText.isEnabled = enabled
            if (customColor) vText.setTextColor(textColor)
        }

        view.setBackgroundColor(background)
    }

    fun onClick(v:View, x:Int, y:Int){
        if (enabled && onClick != null) onClick!!.invoke(v, x, y)
    }

    //
    //  Setters
    //

    fun setVisible(visible: Boolean): CardMenu {
        this.visible = visible
        update()
        return this
    }

    fun setOnClick(onClick: (View, Int, Int) -> Unit): CardMenu {
        this.onClick = onClick
        update()
        return this
    }

    fun setOnLongClick(onLongClick: (View) -> Unit): CardMenu {
        this.onLongClick = onLongClick
        update()
        return this
    }

    fun setDividerVisible(dividerVisible: Boolean): CardMenu {
        this.dividerVisible = dividerVisible
        update()
        return this
    }

    fun setEnabled(enabled: Boolean): CardMenu {
        this.enabled = enabled
        update()
        return this
    }

    fun setIcon(icon: Int): CardMenu {
        this.icon = icon
        update()
        return this
    }

    fun setIcon(icon: Drawable?): CardMenu {
        this.iconDrawable = icon
        update()
        return this
    }

    fun setIconFilter(iconFilter: Int?): CardMenu {
        this.iconFilter = iconFilter
        update()
        return this
    }

    fun setBackgroundRes(@ColorRes background: Int): CardMenu {
        return setBackground(ToolsResources.getColor(background))
    }

    fun setBackground(background: Int): CardMenu {
        this.background = background
        update()
        return this
    }

    fun setText(@StringRes text: Int): CardMenu {
        return setText(ToolsResources.s(text))
    }

    fun setText(text: String?): CardMenu {
        this.text = text ?: ""
        update()
        return this
    }

    fun setChipText(@StringRes chipText: Int): CardMenu {
        return setChipText(ToolsResources.s(chipText))
    }

    fun setChipText(chipText: String?): CardMenu {
        this.chipText = chipText ?: ""
        update()
        return this
    }

    fun setDescription(@StringRes desc: Int): CardMenu {
        return setDescription(ToolsResources.s(desc))
    }

    fun setDescription(desc: String?): CardMenu {
        this.description = desc ?: ""
        update()
        return this
    }

    fun setTextColor(color: Int): CardMenu {
        customColor = true
        textColor = color
        update()
        return this
    }

}